### ISSUE TITLE
Do not always re-read same entries on append 

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/atomix#38]
**Is your feature request related to a problem? Please describe.**

The LeaderAppender holds for each Member an own JournalReader. On sending appendRequests it creates for each Member a new AppendRequest which contains of multiple entries (at max 32 kb). This means for each member the entries are read and copied from the log to send the append request.

This makes sense **if** the followers are not up to date, but in most of the cases they are, which means we spend much time in reading and copying. This also explains my observation which I had with the async profiler that the leaders most of the time spent in reading the log.

It should be possible to use one reader for all followers which are up to date.

**Comment by npepinpe** [comment]
I know Hashicorp's Raft implementation keeps a cache in front of its log for such cases, where followers will be receiving the same entries in the near-future. Of course that brings with it the usual caching issues :sweat_smile: 

I think in general the Journal could be optimized, in my experience it's not very fast unless you do linear reads (i.e. `next()`, `next()`, etc.). Seeking, creating segments, etc., all slow it down, even when using mmap (since it still copies the data anyway by not using direct buffers).

**Comment by Zelldon** [comment]
Yeah I think the implementation is not ideal. We should invest some time in it to improve here. I created a branch with a LRU-Cache will try it on the cluster lets see if we see differences with that

**Comment by npepinpe** [comment]
Ah, yeah just saw. If it helps (which imo I think it will), we could think about using [Caffeine](https://github.com/ben-manes/caffeine) which has size-based eviction, time expiration, metrics, etc. But can be done after to better compare if its worth pulling in another dependency

**Comment by Zelldon** [comment]
Sounds good. Thanks for the suggestion
